docs/learn/learn-comparison-ethereum-2.md

```diff
@@ -124,6 +124,31 @@ as long as the logic compiles to Wasm and the shard provides an "execute block"
 validators. Polkadot has the Substrate development framework that allows full spectrum composability
 with a suite of modules that can be configured, composed, and extended to develop a chain's STF.
 
+## Rollups
+
+Rollups, a layer 2 scaling solution, has been talked about for some time, and their potential to scale Ethereum. A layer 2 network of nodes would be responsible for "rolling up" transactions together by batching them before publishing them to the main Ethereum chain. A rolled-up transaction could include 10,000 transactions. Rollups increase TPS and can reduce transaction fees significantly. When it comes to implementing it, there are a few different ways to approach the problem.
+
+### Optimistic Rollups
+
+Using an "innocent until proven guilty" approach, optimistic rollups will accept transaction commitments to the network while validating for errors and security issues. If invalid, validators can submit fraudulent commitment proof (fraud-proofs). These proofs will be used to reject fraudulent transactions. This way, transaction results along with proof can be stored on the main network, but transaction data can be stored elsewhere. Optimistic rollups can improve transaction scalability 10-100x, and write transactions to Ethereum as [`calldata`](https://ethereum.stackexchange.com/questions/52989/what-is-calldata), reducing gas fees significantly.
+
+Examples of optimistic rollup solutions include [Optimisim](https://www.optimism.io/), [Arbitrum](https://bridge.arbitrum.io/) and [Unipig](https://unipig.exchange/welcome).
+
+### Zero-knowledge Rollups
+
+[Zero-knowledge proofs](https://en.wikipedia.org/wiki/Zero-knowledge_proof) can be used as a "proof" of valid transactions when batching, effectively reducing the transaction payload to be the size of a zk-proof. This, similar to the optimistic approach, allows transaction data to be abstracted away and reduced significantly in size before being committed to the main chain. Similar to most on-chain transactions and states, a ZK-rollup's state is represented as a [merkle tree](https://en.wikipedia.org/wiki/Merkle_tree), and the cryptographic hash of the tree's root(Merkle root) is what is stored on-chain. This allows for efficient batching, and tracking of changes to the transaction states. Any changes to the state will require the operator who initiated the changes to compute a new state root and commit it to the on-chain contract. If the contract verifies the new state as valid, then provided root hash becomes the new state root.
```

Comment:
It's worth mentioning the turing-completeness issue with ZK here.

Reply:
I looked for some good resources online about the completeness problem, but couldn't find a good piece. Let me know if you have any that we can add to the resources.

Reply:
TLD;DR
Expressing arbitrary computation inside a circuit to be proved in zero knowledge is theoretically possible, but slow & cumbersome.

Layer-1:
To reach consensus on a traditional block execution, blockchains rely on re-execution of every transaction. We don't want to allow malicious transactions that cause an endless loop, as it would compromise every miner's/validator's resources and the liveness of the chain. Therefore chains like Ethereum have the concept of gas, so the amount of computation per block is bounded. So even though you might have a notion of Turing completeness, there's actually a hard bound on how many operations you can perform. In addition, as the number of miners/validators scales, the amount of redundant computation scales linearly too, which is bad.

Now on to ZK rollups: 
Good thing is that you avoid the last point, i.e. redundant execution of all transactions. The majority of the network only runs a `verify` operation that should be sublinear in the size of the original Tx's, while the block producer incurs heavy computation to produce the proof.
The challenge for any proof system architect is expressing the (arbitrary) computation as a circuit. For a number of reasons (e.g. all arithmetic needs to be done over a finite field), the circuits that the proof systems "accept" are fairly large and are slower to work with than executing the computation directly (aka natively), requiring up to 1,000,000x work for the prover.
If you want to execute an extremely complicated program, you might run into circuits which are simply too large for the prover to handle, or take too long.


A good article on the limitations of ZK proving systems: https://a16zcrypto.com/measuring-snark-performance-frontends-backends-and-the-future/